src/omit.ts

```diff
@@ -1,49 +1,80 @@
+import { fromPairs } from './fromPairs';
 import { purry } from './purry';
 
 /**
  * Returns a partial copy of an object omitting the keys specified.
- * @param object the object
- * @param names the property names
+ * 
+ * If the properties are optional and they don't exist in the input object then
+ * the same object is returned (unlike when using destructuring, which returns a
+ * new clone of the object).
+
+ * @param data the object
+ * @param propNames the property names
  * @signature
  *    R.omit(obj, names);
  * @example
  *    R.omit({ a: 1, b: 2, c: 3, d: 4 }, ['a', 'd']) // => { b: 2, c: 3 }
+ * 
+ *    const obj: { a?: number, b: number } = { b: 2 };
+ *    const result = R.omit(obj, ['a']);
+ *    result === obj // => true
  * @data_first
  * @category Object
  */
-export function omit<T extends Record<PropertyKey, any>, K extends keyof T>(
-  object: T,
-  names: ReadonlyArray<K>
+export function omit<T extends object, K extends keyof T>(
+  data: T,
+  propNames: ReadonlyArray<K>
 ): Omit<T, K>;
 
 /**
  * Returns a partial copy of an object omitting the keys specified.
- * @param object the object
- * @param names the property names
+ *
+ * If the properties are optional and they don't exist in the input object then
+ * the same object is returned (unlike when using destructuring, which returns a
+ * new clone of the object).
+ *
+ * @param data the object
+ * @param propNames the property names
  * @signature
  *    R.omit(names)(obj);
  * @example
  *    R.pipe({ a: 1, b: 2, c: 3, d: 4 }, R.omit(['a', 'd'])) // => { b: 2, c: 3 }
+ *
+ *    const obj: { a?: number, b: number } = { b: 2 };
+ *    const result = R.pipe(obj, R.omit(['a']));
+ *    result === obj // => true
  * @data_last
  * @category Object
  */
 export function omit<K extends PropertyKey>(
-  names: ReadonlyArray<K>
-): <T extends Record<PropertyKey, any>>(object: T) => Omit<T, K>;
+  propNames: ReadonlyArray<K>
+): <T extends object>(data: T) => Omit<T, K>;
 
 export function omit() {
   return purry(_omit, arguments);
 }
 
-function _omit<T extends Record<PropertyKey, any>, K extends keyof T>(
-  object: T,
-  names: Array<K>
+function _omit<T extends object, K extends keyof T>(
+  data: T,
+  propNames: ReadonlyArray<K>
 ): Omit<T, K> {
-  const set = new Set(names as Array<string>);
-  return Object.entries(object).reduce<any>((acc, [name, value]) => {
-    if (!set.has(name)) {
-      acc[name] = value;
-    }
-    return acc;
-  }, {}) as Omit<T, K>;
+  if (!propNames.some(propName => propName in data)) {
+    // Avoid creating a clone of the input object when the output is unchanged.
+    return data;
+  }
```

Comment:
I'm not sure if returning the actual input is desired. From a user's perspective, I'd want the function to behave consistently - so either always return a new object or the same. Consider:

type FooBar = { foo: string, bar?: number }

const toFoo = (input: FooBar) => omit(input, ['bar'])

const myFooBar: FooBar = { foo: 'foo' }

const myFooOnly = toFoo(myFooBar)
```

here, I'd still expect `myFooOnly` to be a new object, not the same reference as `myFooBar`. If I were to mutate `myFooOnly` later, it should not impact `myFooBar`.

Also, it now behaves differently if I have:

```
const myFooBar2: FooBar = { foo: 'foo', bar: 5 }

const myFooOnly2 = toFoo(myFooBar2)
```

because now, `myFooOnly2` is a new copy because it happened to include the `bar` field ...

Reply:
fair points. I'm thinking maybe we should create some react-friendly util function that would return object A if they are equal, and object B if they are different (via deep equality) so that you could do something like:
```ts
setXXX((current) => {
  return pipe(
    current,
    ...,
    ...,
    ifChanged(current),
  );
});
```

wdyt?

Reply:
yeah that would keep the original reference, and we could do that. additionally (or maybe instead), what would you think about providing something equivalent to the `structuralSharing` feature of react-query (code pointer: https://github.com/TanStack/query/blob/765f6196e900c1dd826b040773fcf97839ca60d0/packages/query-core/src/utils.ts#L217-L250)

it also gives you back the `original` if they are deep equal, however, if they are not, it still tries to keep as many references as possible. For example, if you'd compare & share these two arrays:

```
[{ foo: true }, { bar: false }, { baz: { nested: true } }]
[{ foo: true }, { bar: false }, { baz: { nested: true }, new: 'value' }]
```

You would get back an array with 3 values, where the first and second entry are the same reference as before, and only the last one is a new object. And even there, the `nested` object would also be the same reference as the original one.

Our implementation in Query only works with JSON compatible values though.

Reply:
I think we can start with the simple impl; the difference from react-query being that these are data processing pipes in memory and not fetched from the network and reparsed from scratch, so it's likely that the mutated data would already be a shallow clone.

I just don't want to add too much processing on hot paths, where the whole thing is a perf optimization. The main use-case would be after filters and sorts, where you'd want to prevent a re-render on trivial cases.